### PR TITLE
base64_decode opcode from TEALv7

### PR DESCRIPTION
**Pull request by sczembor**
Closes #x

## Proposed Changes

+ added unit test to test the correctness of the solution
+ added base64_decode opcode functionality 

## TODO

- [x] Add unit tests
- [x] Implement Base64Decode class:
- [x] map the opcode 
- [x] update `CHANGLELOG.MD`
- [x] Add support for tealv7

### CODE REVIEW

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 1]
```diff
@@ -4742,3 +4743,31 @@ export class Gitxnas extends Gtxnas {
...
+ * https://github.com/algorand/go-algorand/pull/3220
+ */
+export class Base64Decode extends Op {
+	readonly line: number;
+	/**
+	 * Asserts 0 arguments are passed.
+	 * @param args Expected arguments: [] // none
+	 * @param line line number in TEAL file
+	 */
+	constructor(args: string[], line: number) {
```
Comment (vuvth) [inline packages/runtime/src/interpreter/opcode-list.ts:4759]: Opcode is `base64_decode e`, so we should have args = [e]. 
e  = {0, 1}.
| Index | Name | Notes |
| - | ------ | --------- |
| 0 | URLEncoding |  |
| 1 | StdEncoding |  |

More details [here](https://github.com/algorand/go-algorand/blob/master/data/transactions/logic/TEAL_opcodes.md#base64_decode-e)

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 2]
```diff
@@ -4742,3 +4743,31 @@ export class Gitxnas extends Gtxnas {
...
+		super();
+		this.line = line;
+		assertLen(args.length, 0, line);
```
Comment (vuvth) [inline packages/runtime/src/interpreter/opcode-list.ts:4762]: Let change it to 1 and assert args[0] is valid value(0 or 1)

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 3]
```diff
@@ -4742,3 +4743,31 @@ export class Gitxnas extends Gtxnas {
...
+	}
+
+	execute(stack: TEALStack): void {
+		this.assertMinStackLen(stack, 1, this.line);
+		const last = this.assertBytes(stack.pop(), this.line);
+		assertBase64(convertToString(last), this.line);
+		const enc = new TextDecoder("utf-8");
```
Comment (vuvth) [inline packages/runtime/src/interpreter/opcode-list.ts:4769]: Change logic to support `URLEncoding` and `StdEncoding`.
I'm also don't know exactly method to do it 😭

File: packages/runtime/src/parser/parser.ts [hunk 4]
```diff
@@ -388,6 +389,14 @@ opCodeMap[6] = {
 	itxnas: ITxnas,
 };
 
+/**
```
Comment (vuvth) [inline packages/runtime/src/parser/parser.ts:392]: Let add test for `base64_decode` in `test/src/parser.ts`

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 5]
```diff
@@ -6612,4 +6613,34 @@ describe("Teal Opcodes", function () {
...
+		const encoded64Base = "Vml0YWxpaw==";
+		const decoded64Base = "Vitalik";
+		const toPush = Buffer.from(encoded64Base, "utf-8");
+		const expectedBytes = new Uint8Array(Buffer.from(decoded64Base, "utf-8"));
+
+		this.beforeEach(() => {
+			stack = new Stack<StackElem>();
+		});
+
+		it("Should decode base64 encoded data and push it to stack", () => {
```
Comment (vuvth) [inline packages/runtime/test/src/interpreter/opcode-list.ts:6628]: Change unit test with new logic.

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 6]
```diff
@@ -6612,4 +6613,34 @@ describe("Teal Opcodes", function () {
...
+			stack.push(toPush);
+			const op = new Base64Decode([], 0);
+			op.execute(stack);
+			assert.deepEqual(expectedBytes, stack.pop());
+		});
+
```
Comment (vuvth) [inline packages/runtime/test/src/interpreter/opcode-list.ts:6634]: Add one more test for invalid argument for opcode.

File: packages/runtime/src/parser/parser.ts [hunk 7]
```diff
@@ -388,6 +389,14 @@ opCodeMap[6] = {
...
+ * TEALv7
+ */
+opCodeMap[7] = {
```
Comment (vuvth) [inline packages/runtime/src/parser/parser.ts:395]: 👍🏻

**Review by vuvth — COMMENTED** [review]
Thanks for your PR! I have few comments. 

How about compute `cost` on this opcode?

Check list: 
- [ ] Update opcode construction and logic.
- [ ] Opcode unit test.
- [ ] Add test for opcode on parser.
- [ ] Compute opcode *cost*. We use it for compute budget of transaction. Logic for compute budget(gas) in method `executeWithResult` of class `runtime/src/interpreter/Interpreter.ts`.

File: packages/runtime/test/src/interpreter/opcode-list.ts (hunk 6 above)
Comment (vuvth) [reply packages/runtime/test/src/interpreter/opcode-list.ts:6634]: maybe you should add above test case in parser.

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -4742,3 +4744,46 @@ export class Gitxnas extends Gtxnas {
...
+			});
+		}
+	}
+
+	execute(stack: TEALStack): void {
+		this.assertMinStackLen(stack, 1, this.line);
+		const last = this.assertBytes(stack.pop(), this.line);
+		const enc = new TextDecoder("utf-8");
+		const decoded = enc.decode(last);
+		if (this.encoding === 0) {
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:4779]: I think we should have an enum for `0/1`

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 9]
```diff
@@ -4742,3 +4744,46 @@ export class Gitxnas extends Gtxnas {
...
+			// UrlEncoding
+			assertBase64Url(convertToString(last), this.line);
+			stack.push(new Uint8Array(Buffer.from(decoded.toString(), "base64url")));
+		} else {
+			// e === 1 StdEncoding
+			assertBase64(convertToString(last), this.line);
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:4785]: you're running the same code in both `if else`. Move it outside

File: packages/runtime/src/lib/constants.ts [hunk 10]
```diff
@@ -322,6 +322,19 @@ export const reOct = /^0[0-8]+$/;
...
+ * ^                          # Start of input
+ * ([0-9a-zA-Z_-]{4})*        # Groups of 4 valid characters decode
+ *                            # to 24 bits of data for each group
+ * (                          # Either ending with:
+ *     ([0-9a-zA-Z_-]{2}==)   # two valid characters followed by ==
+ *     |                      # , or
+ *     ([0-9a-zA-Z_-]{3}=)    # three valid characters followed by =
+ * )?                         # , or nothing
+ * $                          # End of input
+ */
```
Comment (ratik21) [inline packages/runtime/src/lib/constants.ts:335]: Is the text formatting correct here?

File: packages/runtime/test/src/parser/parser.ts [hunk 11]
```diff
@@ -2583,6 +2584,17 @@ describe("Parser", function () {
...
 			assert.deepEqual(res, expected);
 		});
+
+		it("should return correct opcode list for `teal v7`", async () => {
+			const file = "teal-v7.teal";
+			const res = parser(getProgram(file), ExecutionMode.APPLICATION, interpreter);
+			const expected = [
+				new Pragma(["version", "7"], 1, interpreter),
+				new Base64Decode(["0"], 2),
+			];
```
Comment (ratik21) [inline packages/runtime/test/src/parser/parser.ts:2594]: :+1: for adding test in parser as well

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 9 above)
Comment (sczembor) [reply packages/runtime/src/interpreter/opcode-list.ts:4785]: Actually they are different. One is `base64` while the other is `base64Url`

File: packages/runtime/src/lib/constants.ts (hunk 10 above)
Comment (sczembor) [reply packages/runtime/src/lib/constants.ts:335]: It is formatted like that to explain the regex step by step but maybe it is not needed.

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 12]
```diff
@@ -4754,3 +4757,48 @@ export class Gitxnas extends Gtxnas {
...
+	/**
+	 * Asserts 1 argument is passed.
+	 * @param args Expected arguments: [e], where e = {0, 1}.
+	 * @param line line number in TEAL file
+	 */
+	constructor(args: string[], line: number) {
+		super();
+		this.line = line;
+		assertLen(args.length, 1, line);
+		this.encoding = Number(args[0]);
```
Comment (vuvth) [inline packages/runtime/src/interpreter/opcode-list.ts:4778]: Ahh, My bad. The opcode arguments should be {URLEncoding, StdEncoding}.

https://github.com/algorand/go-algorand/blob/5af8ec8506ec770bc3462b2f4f78a284ae22bf94/data/transactions/logic/eval_test.go#L3489

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 13]
```diff
@@ -4754,3 +4757,57 @@ export class Gitxnas extends Gtxnas {
...
+
+	execute(stack: TEALStack): void {
+		this.assertMinStackLen(stack, 1, this.line);
+		const last = this.assertBytes(stack.pop(), this.line);
+		const enc = new TextDecoder("utf-8");
+		const decoded = enc.decode(last);
+		switch (this.encoding) {
+			case Base64Encoding.URL:
+				assertBase64Url(convertToString(last), this.line);
+				stack.push(new Uint8Array(Buffer.from(decoded.toString(), "base64url")));
```
Comment (vuvth) [inline packages/runtime/src/interpreter/opcode-list.ts:4805]: Why we don't store "base64Url" and "base64" to variable? 
We can store it to `this.encoding`. 
Another questions is when I check Opcode class I also see few assert function in Opcode class. 
Should we move `asserBase64URL` and `assertBase64` to Opcode class?

**Review by vuvth — APPROVED** [review]
Nice work! Preapproved!

**Pull request merged**